### PR TITLE
docs: make api params more readable

### PR DESCRIPTION
**Pull request by cunzaizhuyi**
This PR contains:

- [ ] bugfix
- [ ] feature
- [ ] refactor
- [X] documentation
- [ ] other

Are tests included?

- [ ] yes (_bugfixes and features will not be merged without tests_)
- [X] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [X] no

List any relevant issue numbers:

### Description

Rollup api's Type filed write in .md  table, but some api's Type field are very complex, so they look very ugly in .md table. 
I don't know would you like to make them more readable, so i just try to modify one of this api params.

before:
image

after:
image

if you like this, please reply me, i will modify other api.

### CODE REVIEW

**Comment by lukastaegert** [comment]
That is an interesting idea. I was also wondering about something like this, but so far discarded the idea as I saw no nice way to have the types directly in the table. But that way, it would actually work.

A few things that I would want to change, though:
- names: `ResultType` and `ReturnObj` to me are nearly meaningless names. Imagine you add such a section for every single hook. A complete waste of an opportunity. Luckily, those types DO have names in Rollup's own TypeScript types:
  `ReturnObj` is actually `SourceDescription` and `ResultType` is `LoadResult`. Now that would have a double benefit: Each type would have a unique, descriptive name, and TypeScript users would recognize those types when they use them. Also in cases where different hooks share a type, it would really be the SAME type. That would make it much easier also to talk about these types.
- order: When you read the load hook in the new version, the first thing you see now is an unmotivated object definition. Then you see an unmotivated type definition that retroactively explains why we defined that object. And only then you see how it all fits together. I do not think this is a good flow: The eye has to jump all the time, and you are confronted with the details before you even see the high level picture. So why not turn it around completely:

  ### load
  
  |       |                              |
  | ----: | :--------------------------- |
  | Type: | `(id: string) => LoadResult` |
  | ...   |                              |

  ```js
  type LoadResult = string | null | SourceDescription;
  interface SourceDescription {
 	code: string;
 	map?: string | SourceMap;
 	ast?: ESTree.Program;
 	assertions?: { [key: string]: string } | null;
 	meta?: { [plugin: string]: any } | null;
 	moduleSideEffects?: boolean | 'no-treeshake' | null;
 	syntheticNamedExports?: boolean | string | null;
  }
  ```

  That way you would start with the high level picture: Ah, the load hook is a function that gets a string id and returns stuff. Hmmm, I am actually interested in that. Ah, the `LoadResult` can be string, null or a complex object. And only if you are interested in the complex object you continue reading, but you can already put it in the right place in your head. After all, this documentation is not written for computers (which would also have no problem with the reverse order, BTW) but for humans.
 
What do you think?

**Comment by cunzaizhuyi** [comment]
About above two questions:
* If we have all the type definitions, we should use them of course.
* I am completely convinced by your opinion about the human read order. 

so, if we decide to change the form of display of complex types,  how to extract the types?
* where to place these type definitions ? a new document or just put them at the bottom of the same document?

**Comment by lukastaegert** [comment]
That is a good question. Actually, I do not think there are too many. And if the types are small, maybe we can just repeat them at each usage? For bigger types, I currently have one example which is `ChunkInfo` and `AssetInfo`. Those are only defined at the [generateBundle](https://deploy-preview-4852--rollupjs.netlify.app/plugin-development/#generatebundle) hook (very similar to what you proposed actually, the type definition should probably just be moved directly below the table for consistency). But they are referenced in some other places as well like [augmentChunkHash](https://deploy-preview-4852--rollupjs.netlify.app/plugin-development/#augmentchunkhash) which just contains a link to generateBundle. For the banner hook, there we only have a reference to the banner option which again references generateBundle. Not sure if this is the best idea, but then the type is not in a completely arbitrary place but close to its most important usage.

**Comment by cunzaizhuyi** [comment]
OK, let's do it!
please code review.

**Review by lukastaegert — COMMENTED** [review]
Very nice!

Here are some small things I found:

- in `resolveDynamicImport` and `resolveId`, the `PartialResolvedId` should have all of its properties except `id` as `optional` and `| null`.
- Again for `resolveDynamicImport` and `resolveId`, in the original return type I intentionally omitted `| undefined` as `undefined` always has the same effect as returning `null`, but the text always talks about "returning null" and does not explain `undefined`, so it might confuse people. But I am open for discussion as it is a slight deviation from the actual types. This would also apply to `transform`, `renderChunk`, `renderDynamicImport`
- If we do not want to repeat the `ResolveIdResult ` and `PartialResolvedId`, then I would keep them at the `resolveId` hook as that is the more important hook, and write something like "The return type of this hook is the same as that of the [`resolveId`](#resolveid) hook."
- The indentation in `shouldTransformCachedModule` is inconsistent with e.g. `resolveId`. But this is something I would have expected prettier to fix, some maybe just running `npm run lint:markdown` will fix that. Otherwise, maybe you can adjust it manually. This also applies to `renderDynamicImport`
- For consistency, the `AssetInfo` and `ChunkInfo` types in `generateBundle` should probably be interfaces. The code block should probably be moved to the top of the section directly below the table, and in "the full list of files being written or generated along with their details:" the ":" should of course then become a ".".
- Similarly, the types `EmittedChunk` and `EmittedAsset` in `this.emitFile` should be interfaces. The code block should be directly below the table and the line "emittedFile can have one of two forms:" should be placed below the code block as something like "You can emit either chunks or assets."
- And again, the `type ModuleInfo` and `type ResolvedId` in `this.getModuleInfo` should be interfaces. The code block should be directly below the table and the line "Returns additional information about the module in question in the form" should be placed below the code block as "Returns additional information about the module in question."
- In `this.load`, all properties except `id` should be optional and all properties except `id` and `resolveDependencies` should also have an `| null`, cf. the type in the original docs. This is what `Partial<PartialNull<ModuleOptions>>` in the TS types does.
- The return type of `this.resolve` is actually a `ResolvedId`, so we might also just add a link to `this.getModuleInfo` to avoid repeating the type, or alternatively keep the type at `this.resove` and add a link to `this.getModuleInfo`. Or we keep them in two places.

One thought that crossed my mind is: If a type is not directly defined at the hook that uses it but is referenced from another hook, maybe we want to make the type name a link in the type definition? Of course that means we cannot do this in markdown but need some HTML, but for instance for `augmentChunkHash` this could be

Type: <code>(chunkInfo: <a href="#generatebundle">ChunkInfo</a>) => string</code> 

which is written as
`Type: <code>(chunkInfo: <a href="#generatebundle">ChunkInfo</a>) => string</code>`

**Comment by cunzaizhuyi** [comment]
Thank you very much for your point out problems！
I will modify these problems this week.

**Comment by lukastaegert** [comment]
Thank you so much for your effort! It always makes me extremely happy if people help to improve the documentation!

**Comment by cunzaizhuyi** [comment]
@lukastaegert  I modified all the nine problems you said. I admire your earnest and meticulous.

File: docs/plugin-development/index.md [hunk 1]
```diff
@@ -341,11 +355,25 @@ This is the only hook that does not have access to most [plugin context](#plugin
...
+	importer: string,
+	options: { assertions: Record<string, string> }
+) => ResolveIdResult;
+```
+
+::: tip
+
+The return type **ResolveIdResult** is the same as that of the [`resolveId`](#resolveid) hook.
+
+:::
```
Comment (lukastaegert) [inline docs/plugin-development/index.md:375]: Nice idea with the green box, I like that! We are using far too few boxes!

**Review by lukastaegert — APPROVED** [review]
Looks great!
Turns out some of the formatting was slightly inconsistent because of subtle TypeScript issues that prevented prettier from working. I pushed a fix for those.

**Pull request merged**